### PR TITLE
Not retry calling close(...) when EINTR is reported.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Usually we retry the syscall when EINTR is reported. This is not safe in the case of close(...) as the file descriptor may already be closed. The problem arise if this was the case and we retry it and the file descriptor is already reused again. In this case we would close a wrong file descriptor.

See:
  - https://bugs.chromium.org/p/chromium/issues/detail?id=269623
  - https://lwn.net/Articles/576478/

Modifications:

Ignore EINTR on close(...) and just assume the FD was closed.

Result:

Not possible to close the wrong file descriptor when close(...) reports EINTR.

### CODE REVIEW

**Comment by normanmaurer** [comment]
Thanks to @vlm for reporting.

**Comment by normanmaurer** [comment]
@swift-nio-bot test this please

File: Sources/NIO/System.swift [hunk 1]
```diff
@@ -99,14 +99,18 @@ internal func wrapSyscallMayBlock<T: FixedWidthInteger>(where function: StaticSt
...
 }
 
+internal func continueOnEINTR(errnoCode: CInt) -> Bool {
+    return errnoCode == EINTR
+}
+
 /* Sorry, we really try hard to not use underscored attributes. In this case however we seem to break the inlining threshold which makes a system call take twice the time, ie. we need this exception. */
 @inline(__always)
-internal func wrapSyscall<T: FixedWidthInteger>(where function: StaticString = #function, _ body: () throws -> T) throws -> T {
+internal func wrapSyscall<T: FixedWidthInteger>(where function: StaticString = #function, continueOnErrno: (CInt) -> Bool = continueOnEINTR, _ body: () throws -> T) throws -> T {
```
Comment (Lukasa) [inline Sources/NIO/System.swift]: Should we just make this a separate function rather than use the closure?

Comment (normanmaurer) [reply Sources/NIO/System.swift]: @Lukasa I am not strong on it... that said I think the closure should be free so why not just re-use the rest here ? @weissi WDYT ?

Comment (Lukasa) [reply Sources/NIO/System.swift]: I'm a bit reluctant to combine `@inline(__always)` with this closure, that's all.

Comment (normanmaurer) [reply Sources/NIO/System.swift]: ok let me just add an extra method then... Its easy enough to do :)

File: Sources/NIO/System.swift [hunk 2]
```diff
@@ -116,6 +116,19 @@ internal func wrapSyscall<T: FixedWidthInteger>(where function: StaticString = #
...
+@inline(__always)
+internal func wrapSyscallNoRetryOnEINTR<T: FixedWidthInteger>(where function: StaticString = #function, _ body: () throws -> T) throws -> T {
+    let res = try body()
+    if res == -1 {
+        let err = errno
+        assert(!isBlacklistedErrno(err), "blacklisted errno \(err) \(strerror(err)!)")
+        throw IOError(errnoCode: err, function: function)
+    }
+    return res
+}
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply Sources/NIO/System.swift]: works for me as well...

**Comment by normanmaurer** [comment]
@weissi check the links that are included in the comments... Basically there is really nothing we can do about it and the "safest" bet is to assume it worked.

**Comment by normanmaurer** [comment]
@swift-nio-bot test this please

**Comment by normanmaurer** [comment]
@swift-nio-bot test this please

**Comment by vlm** [comment]
@Lukasa it should actually be quite possible to trigger it:
 * Create a TCP *server* that accept a connection with `setsockopt(SO_RCVBUF, 4096)` but doesn't read anything out of it.
 * Create a TCP *client* with `setsockopt(SO_SNDBUF, 4096)` to disable TCP buffers autosizing.
 * Set a non-blocking mode on that client.
 * Attempt to send a lot of data (such as 10⁶ bytes) until EWOULDBLOCK/EAGAIN.
 * Set a blocking mode on a client.
 * Ensure `setsockopt(SO_LINGER)` has not been tampered with.
 * Set up `sigaction(SIGALRM, ~SA_RESTART)` (to disable restarting on an alarm).
 * Set an alarm using `setitimer()`.
 * Invoke `close()`.
 * After alarm goes off the close() should be terminated with -1/EINTR.

I think this should be enough to reliably observe EINTR. If not, the whole thing should be tried with an NFS instead of sockets, but that complicates stuff greatly.

**Comment by vlm** [comment]
No, turned out under Linux it is pretty impossible to get it this way. Take it back.

**Pull request merged**